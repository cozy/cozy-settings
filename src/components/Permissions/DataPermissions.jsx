import React from 'react'

import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import { Link, useParams } from 'react-router-dom'
import withAllLocales from 'lib/withAllLocales'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import NavigationList from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import { routes } from 'constants/routes'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import { retrieveSlugsOfPermission } from './helpers/DataPermissionsHelpers'

import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

const DataPermissions = ({ t }) => {
  const { permission: permissionType } = useParams()
  const THIRTY_SECONDS = 30 * 1000

  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  const slugs = retrieveSlugsOfPermission(
    queryResultApps,
    queryResultKonnectors,
    permissionType
  )

  return (
    <Page narrow>
      <IconButton className="u-mr-half" href={`#${routes.dataList}`}>
        <Icon icon={PreviousIcon} size={16} />
      </IconButton>
      <PageTitle>
        {t('CozyPermissions.Permissions.' + permissionType)}
      </PageTitle>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' &&
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          {slugs.map(slug => {
            return (
              <React.Fragment key={slug.slug}>
                <Link to="" style={{ textDecoration: 'none' }}>
                  <ListItem button>
                    <ListItemIcon>
                      <AppIcon app={slug} />
                    </ListItemIcon>
                    <ListItemText primary={slug.slug} />
                  </ListItem>
                </Link>
                <Divider />
              </React.Fragment>
            )
          })}
        </NavigationList>
      )}
    </Page>
  )
}

export default withAllLocales(DataPermissions)
