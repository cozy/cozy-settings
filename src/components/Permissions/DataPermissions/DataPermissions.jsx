import React from 'react'
import withAllLocales from '../../../lib/withAllLocales'
import { getPermissionsVerbsByType } from '../helpers/permissionsHelper'
import { useParams } from 'react-router-dom'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'
import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import { completePermission } from '../helpers/permissionsHelper'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import PageTitle from 'components/PageTitle'
import {
  displayPermissions,
  getPermissionIconName
} from '../helpers/permissionsHelper'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'

const DataPermissions = ({ t }) => {
  const { permission } = useParams()

  const THIRTY_SECONDS = 30 * 1000
  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  const appsAndKonnectorsSlugs = completePermission(
    queryResultApps,
    queryResultKonnectors
  )[permission]

  const appsAndKonnectors = []

  if (appsAndKonnectorsSlugs?.length > 0) {
    for (let i = 0; i < appsAndKonnectorsSlugs.length; i++) {
      const appToAdd = queryResultApps.data.find(
        app => app.slug === appsAndKonnectorsSlugs[i]
      )
      if (appToAdd) {
        appsAndKonnectors.push(appToAdd)
      }

      const konnectorToAdd = queryResultKonnectors.data.find(
        konnector => konnector.slug === appsAndKonnectorsSlugs[i]
      )
      if (konnectorToAdd) {
        appsAndKonnectors.push(konnectorToAdd)
      }
    }
  }

  const isNotLastItem = slug => {
    return slug !== appsAndKonnectorsSlugs[appsAndKonnectorsSlugs.length - 1]
  }

  const iconName = getPermissionIconName(permission)

  return (
    <Page narrow>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' &&
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <>
          <IconButton className="u-mr-half" href="#/permissions/data">
            <Icon icon={PreviousIcon} size={16} />
          </IconButton>
          <div style={{ textAlign: 'center' }}>
            <Icon
              icon={
                require(`cozy-ui/transpiled/react/Icons/${iconName}`).default
              }
              size={mediumSize}
            />
            <PageTitle>
              {t('CozyPermissions.' + permission).toUpperCase()}
            </PageTitle>
            <Typography variant="body1" className="u-mb-1-half">
              {t('Permissions.access') +
                ' ' +
                t('CozyPermissions.' + permission).toLowerCase()}
            </Typography>
          </div>
          <NavigationList>
            <NavigationListSection>
              {appsAndKonnectors.map(appOrKonnector => {
                return (
                  <div key={appOrKonnector.name}>
                    <ListItem button>
                      <ListItemIcon>
                        <AppIcon app={appOrKonnector} />
                      </ListItemIcon>
                      <ListItemText
                        primary={appOrKonnector.name}
                        secondary={t(
                          displayPermissions(
                            getPermissionsVerbsByType(
                              appOrKonnector.permissions,
                              permission
                            )
                          )
                        )}
                      />
                      <ListItemSecondaryAction>
                        <Icon
                          icon={RightIcon}
                          size={smallSize}
                          className="u-mr-1"
                          style={{ color: 'var(--secondaryTextColor)' }}
                        />
                      </ListItemSecondaryAction>
                    </ListItem>
                    {isNotLastItem(appOrKonnector.slug) && (
                      <Divider variant="inset" />
                    )}
                  </div>
                )
              })}
            </NavigationListSection>
          </NavigationList>
        </>
      )}
    </Page>
  )
}

export default withAllLocales(DataPermissions)
