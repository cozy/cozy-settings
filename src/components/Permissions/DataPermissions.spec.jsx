import { render } from '@testing-library/react'
import React from 'react'
import DataPermissions from './DataPermissions'
import { useQuery } from 'cozy-client'

jest.mock('cozy-ui/transpiled/react/I18n/withLocales', () => {
  return () => Component => {
    const t = text => text
    const match = { params: { permission: 'io.cozy.apps' } }
    return () => <Component match={match} t={t} />
  }
})

jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    Link: ({ narrow, children }) => (
      <div data-testid="page" data-narrow={narrow}>
        {children}
      </div>
    )
  }
})

jest.mock('cozy-client')

jest.mock('components/Page', () => {
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => ({ children }) => (
  <div data-testid="page-title">{children}</div>
))

jest.mock('cozy-ui/transpiled/react/NavigationList', () => {
  return {
    __esModule: true,
    default: ({ children }) => (
      <div data-testid="NavigationList">{children}</div>
    )
  }
})

jest.mock('cozy-ui/transpiled/react/Icon', () => {
  return () => <div data-testid="Icon"></div>
})

jest.mock('cozy-ui/transpiled/react/ListItemText', () => {
  return ({ primary }) => (
    <div data-testid="ListItemText" data-primary={primary}></div>
  )
})

jest.mock('cozy-ui/transpiled/react/AppIcon', () => {
  // eslint-disable-next-line react/display-name
  return () => <div data-testid="AppIcon"></div>
})

describe('DataPermissions', () => {
  beforeEach(() => {
    const queryResultApps = {
      fetchStatus: 'loaded',
      data: [
        {
          permissions: {
            files: {
              type: 'io.cozy.files',
              description: 'Required to access the files'
            },
            allFiles: {
              type: 'io.cozy.files.*',
              description: 'Required to access the files'
            },
            apps: {
              type: 'io.cozy.apps',
              description:
                'Required by the cozy-bar to display the icons of the apps',
              verbs: ['GET']
            }
          },
          slug: 'contacts',
          name: 'Contacts'
        },
        {
          permissions: {
            files: {
              type: 'io.cozy.files',
              description: 'Required to access the files'
            },
            apps: {
              type: 'io.cozy.apps',
              description:
                'Required by the cozy-bar to display the icons of the apps',
              verbs: ['GET']
            }
          },
          slug: 'home',
          name: 'Home'
        },
        {
          permissions: {
            files: {
              type: 'io.cozy.files',
              description: 'Required to access the files'
            }
          },
          slug: 'settings',
          name: 'Settings'
        },
        {
          permissions: {},
          slug: 'store',
          name: 'Store'
        }
      ]
    }
    const queryResultKonnectors = {
      fetchStatus: 'loaded',
      data: [
        {
          permissions: {},
          slug: 'alan',
          name: 'Alan'
        }
      ]
    }
    useQuery.mockReturnValueOnce(queryResultApps)
    useQuery.mockReturnValueOnce(queryResultKonnectors)
  })

  it('should display DataPermissions', () => {
    const { container } = render(<DataPermissions />)
    expect(container).toMatchSnapshot()
  })
})
