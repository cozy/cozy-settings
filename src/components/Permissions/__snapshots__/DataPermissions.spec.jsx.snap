// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPermissions should display DataPermissions 1`] = `
<div>
  <div
    data-narrow="true"
    data-testid="page"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root u-mr-half large"
      href="#/permissions/data"
      tabindex="0"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          data-testid="Icon"
        />
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <div
      data-testid="page-title"
    >
      CozyPermissions.Permissions.undefined
    </div>
    <div
      data-testid="NavigationList"
    />
  </div>
</div>
`;
